src/coordtest/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+//!
+//! The following datadriven directives are supported:
+//! - `sql`: Executes the SQL using transaction rules similar to the
+//! simple pgwire protocol in a new session (that is, multiple `sql`
+//! directives are not in the same session). Output is formatted
+//! [`ExecuteResponse`](coord::ExecuteResponse). The input can contain the
+//! string `<TEMP>` which will be replaced with a temporary directory.
+//! - `wait-sql`: Executes all SQL in a retry loop (with 5s timeout
+//! which will panic) until all datums returned (all columns in all
+//! rows in all statements) are `true`. Prior to each attempt, all
+//! pending feedback messages from the dataflow server are sent to the
+//! Coordinator. Messages for specified items can be skipped by specifying
+//! `exclude-uppers=database.schema.item` as an argument. After each failed
+//! attempt, the timestamp is incremented by 1 to give any new data an
+//! opportunity to be observed.
+//! - `update-upper`: Sends a batch of
+//! [`FrontierUppers`](dataflow::WorkerFeedback::FrontierUppers)
+//! to the Coordinator. Input is one update per line of the format
+//! `database.schema.item N` where N is some numeric timestamp. No output.
+//! - `inc-timestamp`: Increments the timestamp by number in the input. No
+//! output.
+//! - `create-file`: Requires a `name=filename` argument. Creates and truncates
+//! a file with the specified name in the TEMP directory with contents as the
+//! input. No output.
+//! - `append-file`: Same as `create-file`, but appends.
+//! - `print-catalog`: Outputs the catalog. Generally for debugging.
+
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::fs::{File, OpenOptions};
+use std::future::Future;
+use std::io::Write;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+use anyhow::anyhow;
+use futures::future::FutureExt;
+use tempfile::{NamedTempFile, TempDir};
+use tokio::sync::mpsc;
+
+use coord::{
+    session::{EndTransactionAction, Session},
+    Client, ExecuteResponse, SessionClient,
+};
+use dataflow::{WorkerFeedback, WorkerFeedbackWithMeta};
+use dataflow_types::PeekResponse;
+use expr::GlobalId;
+use ore::thread::JoinOnDropHandle;
+use repr::{Datum, Timestamp};
+use timely::progress::change_batch::ChangeBatch;
+
+// The field order matters a lot here so the various threads/tasks are shut
+// down without ever panicing.
+pub struct CoordTest {
+    inner_feedback_tx: mpsc::UnboundedSender<WorkerFeedbackWithMeta>,
+    client: Option<Client>,
+    _handle: JoinOnDropHandle<()>,
+    inner_feedback_rx: mpsc::UnboundedReceiver<WorkerFeedbackWithMeta>,
+    _catalog_file: NamedTempFile,
+    temp_dir: TempDir,
+    uppers: HashMap<GlobalId, Timestamp>,
+    timestamp: Arc<Mutex<u64>>,
+    _verbose: bool,
+}
+
+impl CoordTest {
+    pub async fn new() -> anyhow::Result<Self> {
+        let catalog_file = NamedTempFile::new()?;
+        let (handle, client, inner_feedback_tx, inner_feedback_rx, timestamp) =
```

Comment:
it took me a while to put this together but it seems like

`inner_feedback_rx` is the channel from dataflow workers to the coordtester
and
`inner_feedback_tx` is the channel from the coordtester to the coordinator. is that right. i think that maybe these could be called

`dataflow_feedback_rx` and `coord_feedback_tx` respectively